### PR TITLE
Add build_sif target

### PR DESCRIPTION
**Pull request by karlonw**
Initial Draft

The goal is to allow the Makefile to build sif files based on built targets if BUILD_SIF=1 is passed in on make command line

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -121,9 +135,18 @@ ifneq ($(HPC_LIBS_DIR),)
...
 		.
-ifneq ($(HPC_TMP_LIBS_DIR),)
-	rm -rf $(HPC_LIBS_DIR)
-endif
+        ifneq ($(HPC_TMP_LIBS_DIR),)
+	    rm -rf $(HPC_LIBS_DIR)
+        endif
+        ifeq "$(BUILD_SIF)" "1"
+	    @echo "BUILD_SIF: $(NGC_PYTORCH_HPC_REPO)-ss:$(SHORT_GIT_HASH)"
+	    make build-sif TARGET_NAME="$(NGC_PYTORCH_HPC_REPO)-ss:$(SHORT_GIT_HASH)"
```
Comment (soohoonchoi) [inline Makefile:143]: I'm not sure having ":" in the sif name would be a good idea.  Something that was mentioned in the slack was
ngc-pt-24.10-hpc-<hash> and adding x86/arm64.  Currently it creates sif name like "pytorch-ngc-hpc-dev:d47dc18.sif"

File: Makefile [hunk 2]
```diff
@@ -121,9 +135,18 @@ ifneq ($(HPC_LIBS_DIR),)
...
+        endif
+else
+        ifeq "$(BUILD_SIF)" "1"
+	    @echo "BUILD_SIF: $(NGC_PYTORCH_HPC_REPO):$(SHORT_GIT_HASH)"
+	    make build-sif TARGET_NAME="$(NGC_PYTORCH_HPC_REPO):$(SHORT_GIT_HASH)"
```
Comment (soohoonchoi) [inline Makefile:148]: I'm not sure having ":" in the sif name would be a good idea. Something that was mentioned in the slack was
ngc-pt-24.10-hpc- and adding x86/arm64. Currently it creates sif name like "pytorch-ngc-hpc-dev:d47dc18.sif"

File: Makefile [hunk 3]
```diff
@@ -98,6 +99,19 @@ NGC_PYTORCH_HPC_REPO := pytorch-ngc-hpc-dev
...
 
+# build pytorch sif
+TMP_SIF := $(shell mktemp -d -t sif-reg.XXXXXX)
+TMP_SIF_BASE := "$(PWD)/$(shell basename $(TMP_SIF))"
+.PHONY: build-sif
+build-sif:
+	# Make a tmp dir in the cwd using the tmp_file name.
+	mkdir $(TMP_SIF_BASE)
+	env SINGULARITY_TMPDIR=$(TMP_SIF_BASE) SINGULARITY_CACHEDIR=$(TMP_SIF_BASE) \
+            SINGULARITY_NOHTTPS=true NAMESPACE="" \
```
[the remark this answers is not in the document]
Comment (karlonw) [reply Makefile:118]: If SINGULARITY_TMPDIR or SINGULARITY_CACHEDIR is set, use those, else create CWD tmp versions.  
Does this suffice?

File: Makefile [hunk 4]
```diff
@@ -98,6 +99,19 @@ NGC_PYTORCH_HPC_REPO := pytorch-ngc-hpc-dev
...
+            singularity -vvv build $(TARGET_NAME).sif \
+                             docker-daemon://$(DOCKERHUB_REGISTRY)/$(TARGET_NAME)
```
[the remark this answers is not in the document]
Comment (karlonw) [reply Makefile]: Used docker save and docker-archive as mentioned in README.md.
Is this ok?

**Comment by karlonw** [comment]
Please check that I accurately resolved your concerns above.

**Pull request merged**